internal/corerad/advertise.go

```diff
@@ -521,3 +661,53 @@ func multicastDelay(r *rand.Rand, i int, min, max int64) time.Duration {
 
 	return d
 }
+
+// errLinkNotReady is a sentinel which indicates an interface is not ready
+// for use with an Advertiser.
+var errLinkNotReady = errors.New("link not ready")
+
+// checkInterface verifies the readiness of an interface.
+func checkInterface(iface string) (*net.Interface, error) {
+	// Link must exist.
+	ifi, err := net.InterfaceByName(iface)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get interface %q: %w", ifi.Name, err)
+	}
+
+	// Link must have a MAC address (e.g. WireGuard links do not).
+	if ifi.HardwareAddr == nil {
+		return nil, fmt.Errorf("interface %q has no MAC address", iface)
+	}
+
+	// Link must be up.
+	// TODO: check point-to-point and multicast flags and configure accordingly.
+	if ifi.Flags&net.FlagUp == 0 {
+		return nil, errLinkNotReady
+	}
+
+	// Link must have an IPv6 link-local unicast address.
+	addrs, err := ifi.Addrs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get interface addresses: %w", err)
+	}
+
+	var foundLL bool
+	for _, a := range addrs {
+		// Skip non IP and link-local addresses.
+		a, ok := a.(*net.IPNet)
+		if ok && isIPv6(a.IP) && a.IP.IsLinkLocalUnicast() {
+			foundLL = true
+			break
+		}
+	}
+	if !foundLL {
+		return nil, errLinkNotReady
+	}
+
+	return ifi, nil
+}
+
+// isIPv6 determines if ip is an IPv6 address.
+func isIPv6(ip net.IP) bool {
+	return ip.To16() != nil && ip.To4() == nil
```

Comment:
I usually just check ip.To4() == nil, why the extra To16() check?

Reply:
The former verifies that the address is actually valid:

> To16 converts the IP address ip to a 16-byte representation. If ip is not an IP address (it is the wrong length), To16 returns nil.

While the latter just seems to check for IPv4-ness:

> To4 converts the IPv4 address ip to a 4-byte representation. If ip is not an IPv4 address, To4 returns nil.